### PR TITLE
MbedTLS/TF-M usability improvements for PSA APIs

### LINKED ISSUE #56995: [mbedtls/tf-m] PSA API conflicts
**Describe the bug**
When migrating to Zephyr 3.3.0 which uses mbedtls 3.2.1 and TF-m 1.7.0, we are having PSA API conflicts during build.
After investigation we saw in release note that there is a "PSA API conflicts that can not be immediately resolved between Zephyr's
instance of MbedTLS in the NS environment, and the TF-M PSA APIs" in following commit :
https://github.com/zephyrproject-rtos/zephyr/commit/c63fb217601fb

Does anyone have any roadmap or indications to solve this issue ?

**Expected behavior**
Working as with Zephyr 3.2.0

**Impact**
Currently unable to migrate to Zephyr 3.3.0 but we need it to use TF-m 1.7.0 with new psa_fwu API

**Logs and console output**
Here is an output of our build: 
```
In file included from /home/e.basle/projects/lpp21/lpp4_dl/lpp4_zephyr_3_3_0_integration/platform/zephyr/../modules/crypto/mbedtls/include/mbedtls/ssl_ciphersuites.h:28,
                 from /home/e.basle/projects/lpp21/lpp4_dl/lpp4_zephyr_3_3_0_integration/platform/zephyr/../modules/crypto/mbedtls/include/mbedtls/ssl.h:32,
                 from /home/e.basle/projects/lpp21/lpp4_dl/lpp4_zephyr_3_3_0_integration/framework/base/subsys/common/cert_mngr/lib_cert_mngr/tests_u/src/main_ztest.c:21:
/home/e.basle/projects/lpp21/lpp4_dl/lpp4_zephyr_3_3_0_integration/platform/zephyr/../modules/crypto/mbedtls/include/mbedtls/pk.h:328:36: error: unknown type name 'mbedtls_svc_key_id_t'
  328 |                              const mbedtls_svc_key_id_t key );
      |                                    ^~~~~~~~~~~~~~~~~~~~
/home/e.basle/projects/lpp21/lpp4_dl/lpp4_zephyr_3_3_0_integration/platform/zephyr/../modules/crypto/mbedtls/include/mbedtls/pk.h:994:32: error: unknown type name 'mbedtls_svc_key_id_t'
  994 |                                mbedtls_svc_key_id_t *key,
      |                                ^~~~~~~~~~~~~~~~~~~~
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain Zephyr SDK
 - Zephyr tag 3.3.0

### PR DESCRIPTION
**Pull request by rajkan01**
## Linked PRs

- https://github.com/zephyrproject-rtos/mbedtls/pull/43
- https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/92

## Description

Previously, Zephyr's mbedtls module's cmake build created a single static library, rather than the collection of libraries (`mbedtls`, `mbedcrypto`, and `mbedx509`) that upstream mbedTLS cmake provides. 

To give better control at link time to choose the required libraries to link in, this commit updates the Zephyr MbedTLS module to also define three distinct libraries rather than a single static MbedTLS library.

One benefit of the three library approach is that if mbedTLS is used in Zephyr in the the non-secure application in addition to TFM's PSA Crypto API on the secure side with TF-M, PSA API calls on the non-secure side will be redirected to the TFM PSA implementation, and the `mbedcrypto` library will only be linked to the secure (TF-M) binary, with the `mbedtls` and `mbedx509` libraries linked against the non-secure (Zephyr) binary, enabling TLS calls to PSA crypto to be redirected to `mbedcrypto` in the secure partition and avoiding function duplication in the non-secure binary.

## Additional Changes

Re-commits the `psa_crypto` sample, which had to be removed during the TF-M 1.7.0 update due to unresolvable linker issues with MbedTLS as a single static library.

Fixes #56995

### CODE REVIEW

**Comment by ceolin** [comment]
[1c1dd384b877bf6b3fa66bfb9ba85c2528994c5d](https://github.com/zephyrproject-rtos/zephyr/commit/1c1dd384b877bf6b3fa66bfb9ba85c2528994c5d) -> can we make this include path automatic when `CONFIG_MBEDTLS` is selected ? That is breaking the current usage and and it is not the common pattern on Zephyr AFAIR.

Other than that looks good to me.

**Comment by microbuilder** [comment]
> [1c1dd384b877bf6b3fa66bfb9ba85c2528994c5d](https://github.com/zephyrproject-rtos/zephyr/commit/1c1dd384b877bf6b3fa66bfb9ba85c2528994c5d) -> can we make this include path automatic when `CONFIG_MBEDTLS` is selected ? That is breaking the current usage and and it is not the common pattern on Zephyr AFAIR.
> 
> Other than that looks good to me.

Thanks for the feedback, Flavio. There is a precedent for this here, for example: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/sockets/CMakeLists.txt#L35 ... but can you point to an example of what you'd like to see here instead?

File: modules/mbedtls/CMakeLists.txt
Comment (rajkan01) [reply modules/mbedtls/CMakeLists.txt:14]: I want to expose the mbedTLS headers whoever target_link_libraries with mbedTLS so created as the interface libraries, however, the zephyr extension zephyr_library_include_directories same function calls the target_include_directories indirectly with PRIVATE to prevent exposing the headers and same issue with zephyr_library_compile_definitions.

Because of that using the direct target_include_directories and target_compile_definitions CMake calls, so called CMake function APIs for creating the mbedTLS headers interface library rather than using zephyr extension functions.

File: modules/mbedtls/CMakeLists.txt
Comment (nordicjm) [inline modules/mbedtls/CMakeLists.txt:29]: Why has INTERFACE been moved to a separate line for these?

File: modules/mbedtls/CMakeLists.txt
Comment (nordicjm) [inline modules/mbedtls/CMakeLists.txt:39]: Why is this being added to a list instead of directly provided to zephyr_library_sources if its only use is being provided to zephyr_library_sources?

File: modules/mbedtls/CMakeLists.txt
Comment (nordicjm) [inline modules/mbedtls/CMakeLists.txt:175]: As above - why?

File: modules/mbedtls/CMakeLists.txt
Comment (microbuilder) [reply modules/mbedtls/CMakeLists.txt:14]: > What is not working when using the zephyr functions?

It sounds like due to the `PRIVATE` flag in the zephyr extensions, the header files are not being exported, but @rajkan01 can hopefully clarify further.

File: tests/subsys/jwt/CMakeLists.txt
Comment (joerchan) [inline tests/subsys/jwt/CMakeLists.txt:13]: ifdef(CONFIG_MBEDTLS_BUILTIN)
The library is under that selection.

File: tests/crypto/mbedtls/CMakeLists.txt
Comment (joerchan) [inline tests/crypto/mbedtls/CMakeLists.txt:11]: CONFIG_MBEDTLS_BUILTIN

File: tests/benchmarks/mbedtls/CMakeLists.txt
Comment (joerchan) [inline tests/benchmarks/mbedtls/CMakeLists.txt:11]: CONFIG_MBEDTLS_BUILTIN

File: modules/mbedtls/CMakeLists.txt
Comment (rajkan01) [reply modules/mbedtls/CMakeLists.txt:14]: yes, it is a problem with the PRIVATE flag in the zephyr extensions.

File: modules/mbedtls/CMakeLists.txt
Comment (rajkan01) [reply modules/mbedtls/CMakeLists.txt]: I fixed the alignment issue.

File: modules/mbedtls/CMakeLists.txt
Comment (rajkan01) [reply modules/mbedtls/CMakeLists.txt:37]: Make the CMake more readable and give a clear source segregation base name as all are from the same library directory.
Please let me know, is any particular reason you want to add under directly to zephyr_library_sources

File: modules/mbedtls/CMakeLists.txt
Comment (rajkan01) [reply modules/mbedtls/CMakeLists.txt:173]: https://github.com/zephyrproject-rtos/zephyr/pull/58023#discussion_r1202135642

File: modules/mbedtls/CMakeLists.txt
Comment (nordicjm) [reply modules/mbedtls/CMakeLists.txt:14]: Can you give an example because I switched in this PR to use the old method and build smp_svr with sha256 support and it built fine?

Comment (microbuilder) [reply modules/mbedtls/CMakeLists.txt:14]: You might want to try the `psa_crypto` sample in this PR, since that brings out the problems we're trying to solve with MbedTLS. Some variation of this (QEMU used below):

```bash
$ west build -p auto -b mps2_an521_ns path/psa_crypto -t run
```

Comment (nordicjm) [reply modules/mbedtls/CMakeLists.txt:14]: > You might want to try the psa_crypto sample in this PR, since that brings out the problems we're trying to solve with MbedTLS. Some variation of this (QEMU used below):

Tried that using the old `zephyr_interface_library_named(mbedTLS)` style with this PR and it builds fine.

> I think no, if the particular application or other libraries depend on mbedTLS, they should explicitly link libraries with mbedTLS as other does in the Zephyr source tree.

But that's no different than it was and using the zephyr_ functions is no different in that regard?

> Another problem, I could not find a way to overcome the issue of automatically adding header inclusion using Zephyr CMake, if you have suggestions to fix this, that would be good.

As in adjusting the include paths for the zephyr application as a whole?

**Comment by microbuilder** [comment]
@joerchan @nordicjm Mind having another look to try to get this in before the freeze Friday? Thanks.

**Review by ceolin — DISMISSED** [review]

File: modules/mbedtls/CMakeLists.txt
Comment (nordicjm) [inline modules/mbedtls/CMakeLists.txt:14]: This should be moved up to where it was previously so that it's included with the else() below as it was previously

**Review by nordicjm — CHANGES_REQUESTED** [review]
Will +1 after this change

**Review by nordicjm — DISMISSED** [review]

**Review by joerchan — CHANGES_REQUESTED** [review]
I want to be stricter in what are exposed in interface include paths.

sha512.h could be define by other modules for example.
mbedtls/sh512.h is very clear.

**Comment by rajkan01** [comment]
> I want to be stricter in what are exposed in interface include paths.
> 
> sha512.h could be define by other modules for example. mbedtls/sh512.h is very clear.

I would say this is already taken care of in the mbedTLS sha512.h file if the user defines this flag: https://github.com/zephyrproject-rtos/mbedtls/blob/zephyr/include/mbedtls/sha512.h#L40.

Please let me know if there is any other mbedTLS header file that may have a problem like that.

File: modules/mbedtls/CMakeLists.txt
Comment (rajkan01) [reply modules/mbedtls/CMakeLists.txt]: Sorry, I removed those include/mbedtls and include/psa as I added them initially and later did not try to remove and test the psa_crypto to find out the dependency.

**Review by ceolin — APPROVED** [review]

**Review by ceolin — DISMISSED** [review]

**Review by d3zd3z — DISMISSED** [review]
The changes look good, but we'll have to wait for the TF-M and mbedtls changes to merge, and update the reference. I'll approve, as there appears to be a DNM on it.

**Comment by joerchan** [comment]
> > I want to be stricter in what are exposed in interface include paths.
> > sha512.h could be define by other modules for example. mbedtls/sh512.h is very clear.
> 
> @ [joerchan](https://github.com/joerchan) I would say this is already taken care of in the mbedTLS sha512.h file if the user defines this flag: https://github.com/zephyrproject-rtos/mbedtls/blob/zephyr/include/mbedtls/sha512.h#L40.
> 
> Please let me know if there is any other mbedTLS header file that may have a problem like that.

You misunderstood me, I'm not talking about mbedtls ALT headers for sh512. I'm talking about other modules having their own sh512 header.

In zephyr you can include:
<tinycrypt/sha512>
<mbedtls/sha512>
If you include:
<sha512.h>
Then which one would you get?

The hostap module (Wi-FI) for example has a sha512.h header file that will be included like "sha512.h". Now I have to make sure that correct sha header file is included by looking at header file path order.

So I don't want "include/mbedtls" on the include path for this reason.

**Comment by rajkan01** [comment]
> > > I want to be stricter in what are exposed in interface include paths.
> > > sha512.h could be define by other modules for example. mbedtls/sh512.h is very clear.
> > 
> > 
> > @ [joerchan](https://github.com/joerchan) I would say this is already taken care of in the mbedTLS sha512.h file if the user defines this flag: https://github.com/zephyrproject-rtos/mbedtls/blob/zephyr/include/mbedtls/sha512.h#L40.
> > Please let me know if there is any other mbedTLS header file that may have a problem like that.
> 
> You misunderstood me, I'm not talking about mbedtls ALT headers for sh512. I'm talking about other modules having their own sh512 header.
> 
> In zephyr you can include: <tinycrypt/sha512.h> <mbedtls/sha512.h> If you include: <sha512.h> Then which one would you get?
> 
> The hostap module (Wi-FI) for example has a sha512.h header file that will be included like "sha512.h". Now I have to make sure that correct sha header file is included by looking at header file path order.
> 
> So I don't want "include/mbedtls" and "include/psa" on the include path for this reason.

Sorry, as you said, I misunderstood. I made the library include path changes pushed here already; please review and approve if you are okay with the final changes.

**Review by joerchan — DISMISSED** [review]

**Comment by joerchan** [comment]
Thanks @rajkan01  Looks good to me now.

**Review by d3zd3z — APPROVED** [review]
Looks good, as long as CI passes.

**Review by microbuilder — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Pull request merged**
